db/map.go

```diff
@@ -0,0 +1,319 @@
+	insertTableRegexp = regexp.MustCompile(`(?i)^\s*insert\s+into\s+([a-z\d \_,` + "`" + `]+)\s+(?:set|\()`)
+	// updateTableRegexp matches the table name from an SQL update statement
+	updateTableRegexp = regexp.MustCompile(`(?i)^\s*update\s+([a-z\d \_,` + "`" + `]+)\s+set`)
+	// deleteTableRegexp matches the table name from an SQL delete statement
+	deleteTableRegexp = regexp.MustCompile(`(?i)^\s*delete\s+from\s+([a-z\d \_,` + "`" + `]+)\s+where`)
+
+	// tableRegexps is a list of regexps that tableFromQuery will try to use in
+	// succession to find the table name for an SQL query.
+	tableRegexps = []*regexp.Regexp{
+		selectTableRegexp,
+		insertTableRegexp,
+		updateTableRegexp,
+		deleteTableRegexp,
+	}
+)
+
+// tableFromQuery uses the tableRegexps on the provided query to return the
+// associated table name or an empty string if it can't be determined from the
+// query.
+func tableFromQuery(query string) string {
+	for _, r := range tableRegexps {
+		if matches := r.FindStringSubmatch(query); len(matches) >= 2 {
+			return matches[1]
+		}
+	}
+	return ""
+}
+
+// WrappedExecutor wraps a gorp.SqlExecutor such that its major functions
+// wrap error results in ErrDatabaseOp instances before returning them to the
+// caller.
+type WrappedExecutor struct {
+	gorp.SqlExecutor
+}
+
+func (we WrappedExecutor) Get(holder interface{}, keys ...interface{}) (interface{}, error) {
+	res, err := we.SqlExecutor.Get(holder, keys...)
+	if err != nil {
+		return res, ErrDatabaseOp{
+			Op:    "get",
+			Table: fmt.Sprintf("%T", holder),
+			Err:   err,
+		}
+	}
+	return res, err
+}
+
+func (we WrappedExecutor) Insert(list ...interface{}) error {
+	if err := we.SqlExecutor.Insert(list...); err != nil {
+		table := "unknown"
+		if len(list) > 0 {
+			table = fmt.Sprintf("%T", list[0])
+		}
+		return ErrDatabaseOp{
+			Op:    "insert",
+			Table: table,
+			Err:   err,
+		}
+	}
+	return nil
+}
+
+func (we WrappedExecutor) Update(list ...interface{}) (int64, error) {
+	updatedRows, err := we.SqlExecutor.Update(list...)
+	if err != nil {
+		table := "unknown"
+		if len(list) > 0 {
+			table = fmt.Sprintf("%T", list[0])
+		}
+		return updatedRows, ErrDatabaseOp{
+			Op:    "update",
+			Table: table,
+			Err:   err,
+		}
+	}
+	return updatedRows, err
+}
+
+func (we WrappedExecutor) Delete(list ...interface{}) (int64, error) {
+	deletedRows, err := we.SqlExecutor.Delete(list...)
+	if err != nil {
+		table := "unknown"
+		if len(list) > 0 {
+			table = fmt.Sprintf("%T", list[0])
+		}
+		return deletedRows, ErrDatabaseOp{
+			Op:    "delete",
+			Table: table,
+			Err:   err,
+		}
+	}
+	return deletedRows, err
+}
+
+func (we WrappedExecutor) Select(holder interface{}, query string, args ...interface{}) ([]interface{}, error) {
+	result, err := we.SqlExecutor.Select(holder, query, args...)
+	if err != nil {
+		table := fmt.Sprintf("%T", holder)
+		if extractedTable := tableFromQuery(query); extractedTable != "" {
```

Comment:
Since we expect in most cases to be able to extract the table, I think it makes more sense to put this function (and similar ones) in order by expected flow:

```
table := tableFromQuery(query)
if table == "" {
  table := fmt.Sprintf("%T", holder)
}
```

Reply:
Related: when we can't explicitly derive the table name we should probably indicate that in the error, i.e. `fmt.Sprintf("%T (unknown table)", holder)`.